rfc/parameterizedMetrics.md

```diff
@@ -0,0 +1,124 @@
+
+So
+```
+"metricName(parameterName=parameterValue)"
+```
+
+will result in a dynamic metric based on `metricName`, but substituting the parameter value when calculating it.  Also, the above string will be used as the column name in the resulting report response schema. 
+
+Motivation
+----------
+
+Standard metrics in Fili are configured at load time.  However if reporting requires many different similar caluclations, such as adjusting precision or converting values of different units into a shared aggregate, or applying custom metric filters, it may be impractical or even impossible to configure all the interesting report metrics.
+
+With parameterized metrics, open-ended variations may be supported for a given conceptual metric calculation, simplifying configuration and possibly unlocking otherwise impractical freedom in metric calculation.
+
+
+The structure of a parameterized metric:
+----------------------------------------
+
+```
+metricName(parameterName=value)
+```
+
+metricName -> an api metric name which should appear as a resource at the `/v1/metrics/metricName` endpoint.
+
+Balanced outer parentheses are the signal that a metric is to be processed as parameterized metric.  Imbalanced parentheses or unescaped internal parentheses will probably result in request errors.
+
+parameterName -> an identifier for a parameter supported by this metric.  Identifier names should follow the same format rules as api resource identifiers.
+
+value -> a value to be used for creating a modified version of baseMetric.  URL encoding may be required if some of the characters in the value contain special characters.  Escaping values containing ampersand is mandatory and it may also be necessary to encode values containing comma or parentheses depending on implementation.
+
+If multiple parameter names are used for this dimension they should be joined by commas with no intermediate spaces (similar to the filter value syntax)
+
+i.e.
+
+```
+metricName(parameterName=value,otherParameter=otherValue)
+```
+
+Response schema
+---------------
+
+Unless aliasing is supported and used (see below), responses will use the full string of the metric expression as the column name for values resulting from parametrized metric evaluation.
+
+
+Reserved parameter names
+------------------------
+
+Two optional features which may be supported in some fili implementations are 'filter' and 'AS'.  These parameter names should be treated as reserved when configuring metrics.
+
+Dynamic aggregation filters, if supported, use the parameter name `filter`. Values will be parsed according to the same rules as the entries in the `filters` query parameter.  The expected behavior of the filter parameter is that rows in the base fact will not be aggregated unless the included filter is true.  This is potentially costly in performance and global filters should always be preferred to dynamic filters, if possible.
```

Comment:
We should be consistent with our reserved keywords, either all lowercase or all uppercase. I personally vote for all lowercase, because that's slightly easier to type.

Reply:
I also think we should use `filters` rather than `filter`, for the sake of consistency.

Reply:
the `filters` clause takes a list of filter expressions.  A single `filter` is what the filter clause would take.  Or at least that's what I was originally thinking.

I suppose there's no reason it couldn't take a list of filters.  I'll update accordingly.

Reply:
This actually suggests a larger pattern that could be followed.  Specifically, we could have other query limiting terms be reserved.  reserve 'dateTime' and provide a default time based filter on aggregation.

Reply:
Actually, there's a noisome problem, which is that the filters clause supports internal commas.  Which will conflict with the commas used to separate fields within the parameter list.  I'm not sure how to readably resolve that.

Reply:
Speaking of case sensitivity, sounds like you are suggesting parameter names are case sensitive. Can we call that out.

Reply:
I think singular (filter) makes more sense here that plural (filters), even though the filter expression can be composed of multiple components that are AND'd together. Logically, I like druid's model, where filter is always singular.

Reply:
singular filter aligns with a tree expression.  If we ever support trees of constraints then we can support singlular filter.